db/schema_tables.cc

```diff
@@ -1095,26 +1166,88 @@ future<> store_column_mapping(distributed<service::storage_proxy>& proxy, schema
 // and or filesystem calls, e.g. fsync.
 constexpr size_t max_concurrent = 8;
 
+// Extracts the names of tables affected by a schema mutation.
+// The mutation must target one of the tables in schema_tables_holding_schema_mutations().
+static
+table_selector get_affected_tables(const sstring& keyspace_name, const mutation& m) {
+    const schema& s = *m.schema();
+    auto get_table_name = [&] (const clustering_key& ck) {
+        // The first component of the clustering key in each table listed in
+        // schema_tables_holding_schema_mutations contains the table name.
+        return value_cast<sstring>(utf8_type->deserialize(ck.get_component(s, 0)));
+    };
+    table_selector result;
+    if (m.partition().partition_tombstone()) {
+        slogger.trace("Mutation of {}.{} for keyspace {} contains a partition tombstone",
+                      m.schema()->ks_name(), m.schema()->cf_name(), keyspace_name);
+        result.all_in_keyspace = true;
+    }
+    for (auto&& e : m.partition().row_tombstones()) {
+        const range_tombstone& rt = e.tombstone();
+        if (rt.start.size(s) == 0 || rt.end.size(s) == 0) {
+            slogger.trace("Mutation of {}.{} for keyspace {} contains a multi-table range tombstone",
+                          m.schema()->ks_name(), m.schema()->cf_name(), keyspace_name);
+            result.all_in_keyspace = true;
+            break;
+        }
+        auto table_name = get_table_name(rt.start);
+        if (table_name != get_table_name(rt.end)) {
+            slogger.trace("Mutation of {}.{} for keyspace {} contains a multi-table range tombstone",
+                          m.schema()->ks_name(), m.schema()->cf_name(), keyspace_name);
+            result.all_in_keyspace = true;
+            break;
+        }
+        result.add(table_name);
+    }
+    for (auto&& row : m.partition().clustered_rows()) {
+        result.add(get_table_name(row.key()));
+    }
+    slogger.trace("Mutation of {}.{} for keyspace {} affects tables: {}, all_in_keyspace: {}",
+                  m.schema()->ks_name(), m.schema()->cf_name(), keyspace_name, result.tables, result.all_in_keyspace);
+    return result;
+}
+
 static future<> do_merge_schema(distributed<service::storage_proxy>& proxy, std::vector<mutation> mutations, bool do_flush)
 {
     slogger.trace("do_merge_schema: {}", mutations);
     schema_ptr s = keyspaces();
     // compare before/after schemas of the affected keyspaces only
     std::set<sstring> keyspaces;
     std::set<table_id> column_families;
+    std::unordered_map<keyspace_name, table_selector> affected_tables;
     for (auto&& mutation : mutations) {
-        keyspaces.emplace(value_cast<sstring>(utf8_type->deserialize(mutation.key().get_component(*s, 0))));
+        sstring keyspace_name = value_cast<sstring>(utf8_type->deserialize(mutation.key().get_component(*s, 0)));
+
+        if (schema_tables_holding_schema_mutations().contains(mutation.schema()->id())) {
```

Comment:
Do we expect mutations not relevant to the schema here?

Reply:
"schema mutations" here refers to the `schema_mutations` object.